File: C#/PlatformodePaymentIntegration/PaySmart2D.cs
using PlatformodePaymentIntegration.Contract.Request;
using PlatformodePaymentIntegration.Contract.Response;
using PlatformodePaymentIntegration.Extension;
using PlatformodePaymentIntegration.Generate;
using PlatformodePaymentIntegration.Settings;
using System.Text.Json;

namespace PlatformodePaymentIntegration;

public class PaySmart2D
{
    private const string URL = "api/paySmart2D";

    private readonly HttpClient _httpClient;
    private readonly ApiSettings _apiSettings;

    public PaySmart2D()
    {
        _httpClient = new HttpClient();
        _apiSettings = new ApiSettingConfiguration().Configuration();
    }

    private async Task<PaySmart2DResponse?> GetAsync()
    {
        var tokenResponse = await new TokenApi().GetAsync();

        if (tokenResponse == null)
        {
            throw new ArgumentNullException("Token bilgisi alınamadı. Lütfen appsettings.json dosyasındaki bilgileri kontrol ediniz.");
        }

        PaySmart2DRequest paySmart2DRequest = CreateRequestParameter(_apiSettings);

        var jsonRequest = JsonSerializer.Serialize(paySmart2DRequest);

        var httpContent = new StringContent(jsonRequest, System.Text.Encoding.UTF8, "application/json");

        _httpClient.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", tokenResponse.data.token);

        var httpResponse = await _httpClient.PostAsync($"{_apiSettings.BaseAddress}{URL}", httpContent);

        var jsonResponse = await httpResponse.Content.ReadAsStringAsync();

        return JsonSerializer.Deserialize<PaySmart2DResponse>(jsonResponse);
    }

    private PaySmart2DRequest CreateRequestParameter(ApiSettings apiSettings)
    {
        PaySmart2DRequest paySmart2DRequest = new()
        {
            cc_holder_name = "Test kart",
            cc_no = "4132260000000003",
            expiry_month = "12",
            expiry_year = "2024",
            cvv = "555",
            currency_code = "TRY",
            installments_number = 1,
            invoice_id = InvoiceGenerator.GenerateInvoiceId(),
            invoice_description = "INVOICE TEST DESCRIPTION",
            total = 10,
            items = new List<Item2D>
            {
                new Item2D()
                {
                    name = "item",
                    price = 10,
                    quantity = 1,
                    description = "item description"
                }
            },
            name = "John",
            surname = "Dao",
            merchant_key = apiSettings.MerchantKey
        };

        HashGenerator hashGenerator = new();

        paySmart2DRequest.hash_key = hashGenerator.GenerateHashKey(
            false,
            paySmart2DRequest.total.ToString(),
            paySmart2DRequest.installments_number.ToString(),
            paySmart2DRequest.currency_code,
            paySmart2DRequest.merchant_key,
            paySmart2DRequest.invoice_id);

        return paySmart2DRequest;
    }

    public async Task PrintAsync()
    {
        PaySmart2DRequest paySmart2DRequest = CreateRequestParameter(_apiSettings);

        var response = await GetAsync();

        Console.WriteLine();
        ConsoleExtensions.BoxedOutput("Endpoint Bilgileri");
        ConsoleExtensions.WriteLineWithSubTitle("Endpoint Adresi : ", _apiSettings.BaseAddress + URL);
        ConsoleExtensions.WriteLineWithSubTitle("Method Tipi : ", HttpMethod.Post);
        ConsoleExtensions.BoxedOutput("Request Parametreler");
        ConsoleExtensions.WriteLineWithSubTitle($"cc_holder_name : ", paySmart2DRequest.cc_holder_name);
        ConsoleExtensions.WriteLineWithSubTitle($"cc_no : ", paySmart2DRequest.cc_no);
        ConsoleExtensions.WriteLineWithSubTitle($"expiry_month : ", paySmart2DRequest.expiry_month);
        ConsoleExtensions.WriteLineWithSubTitle($"expiry_year : ", paySmart2DRequest.expiry_year);
        ConsoleExtensions.WriteLineWithSubTitle($"cvv : ", paySmart2DRequest.cvv);
        ConsoleExtensions.WriteLineWithSubTitle($"currency_code : ", paySmart2DRequest.currency_code);
        ConsoleExtensions.WriteLineWithSubTitle($"installments_number : ", paySmart2DRequest.installments_number);
        ConsoleExtensions.WriteLineWithSubTitle($"invoice_id : ", paySmart2DRequest.invoice_id);
        ConsoleExtensions.WriteLineWithSubTitle($"invoice_description : ", paySmart2DRequest.invoice_description);
        ConsoleExtensions.WriteLineWithSubTitle($"total : ", paySmart2DRequest.total);
        ConsoleExtensions.WriteLineWithSubTitle($"items : ", JsonSerializer.Serialize(paySmart2DRequest.items));
        ConsoleExtensions.WriteLineWithSubTitle($"name : ", paySmart2DRequest.name);
        ConsoleExtensions.WriteLineWithSubTitle($"surname : ", paySmart2DRequest.surname);
        ConsoleExtensions.WriteLineWithSubTitle($"merchant_key : ", paySmart2DRequest.merchant_key);
        ConsoleExtensions.WriteLineWithSubTitle($"hash_key : ", paySmart2DRequest.hash_key);
        ConsoleExtensions.BoxedOutput("Response Parametreler");

        if (response != null)
        {
            ConsoleExtensions.WriteLineWithSubTitle($"status_code : ", response.status_code);
            ConsoleExtensions.WriteLineWithSubTitle($"status_description : ", response.status_description);
            ConsoleExtensions.WriteLineWithSubTitle($"order_no : ", response.data?.order_no);
            ConsoleExtensions.WriteLineWithSubTitle($"order_id : ", response.data?.order_id);
            ConsoleExtensions.WriteLineWithSubTitle($"invoice_id : ", response.data?.invoice_id);
            ConsoleExtensions.WriteLineWithSubTitle($"credit_card_no : ", response.data?.credit_card_no);
            ConsoleExtensions.WriteLineWithSubTitle($"transaction_type : ", response.data?.transaction_type);
            ConsoleExtensions.WriteLineWithSubTitle($"payment_status : ", response.data?.payment_status);
            ConsoleExtensions.WriteLineWithSubTitle($"payment_method : ", response.data?.payment_method);
            ConsoleExtensions.WriteLineWithSubTitle($"error_code : ", response.data?.error_code);
            ConsoleExtensions.WriteLineWithSubTitle($"error : ", response.data?.error);
            ConsoleExtensions.WriteLineWithSubTitle($"auth_code : ", response.data?.auth_code);
            ConsoleExtensions.WriteLineWithSubTitle($"merchant_commission : ", response.data?.merchant_commission);
            ConsoleExtensions.WriteLineWithSubTitle($"user_commission : ", response.data?.user_commission);
            ConsoleExtensions.WriteLineWithSubTitle($"merchant_commission_percentage : ", response.data?.merchant_commission_percentage);
            ConsoleExtensions.WriteLineWithSubTitle($"merchant_commission_fixed : ", response.data?.merchant_commission_fixed);
            ConsoleExtensions.WriteLineWithSubTitle($"hash_key : ", response.data?.hash_key);
            ConsoleExtensions.WriteLineWithSubTitle($"original_bank_error_code : ", response.data?.original_bank_error_code);
            ConsoleExtensions.WriteLineWithSubTitle($"original_bank_error_description : ", response.data?.original_bank_error_description);
        }
        else
        {
            Console.WriteLine("Response alınamadı.");
        }
    }
}

File: PlatformPHP/Paysmart2D.php
<?php

$baseUrl = "https://testapp.platformode.com.tr/ccpayment/api/paySmart2D";

// Token Request
$app_id = "de948c3eafdf5582409d0ad9a0809666";

// Token & 2D Request
$appSecret = "b15fba89a18997ab32e36d0b490f9aff	";

// 2D Request
$total = 110.00;
$installment = 1;
$merchantKey = '$2y$10$avMpLZvIIEY4brcULaj4u.can9eg3gAnx5s3JGz5Yxd.9zka8YfaO';
$currencyCode = "TRY";
$invoice_id = "Test9861";


$tokenResponse = getToken($app_id, $appSecret);

$decodedTokenResponse = json_decode($tokenResponse, true);

if ($decodedTokenResponse['status_code'] == 100) {
    $token = $decodedTokenResponse['data']['token'];
} else {
    echo "Token alınamadı. Lütfen bilgilerinizi kontrol ediniz.";
	return;
}

$data = array(
    "cc_holder_name" => "John Dao",
    "cc_no" => "4155141122223339",
    "expiry_month" => "12",
    "expiry_year" => "2025",
    "cvv" => "555",
    "currency_code" => $currencyCode,
    "installments_number" => $installment,
    "invoice_id" => $invoice_id,
    "invoice_description" => "INVOICE TEST DESCRIPTION",
    "total" => $total,
    "merchant_key" => $merchantKey,
    "items" => array(
        array(
            "name" => "item",
            "price" => 110.00,
            "quantity" => 1,
            "description" => "item description"
        )
    ),
    "name" => "John",
    "surname" => "Dao",
    "payment_status" => 1,
    "transaction_type" => "Auth",
    "hash_key" => generateHashKey($total, $installment, $currencyCode, $merchantKey, $invoice_id, $appSecret)
);

$ch = curl_init($baseUrl);

curl_setopt($ch, CURLOPT_HTTPHEADER, array(
    'Content-Type: application/json',
    "Authorization: Bearer $token"
));

$jsonData = json_encode($data);

curl_setopt($ch, CURLOPT_POST, true);
curl_setopt($ch, CURLOPT_POSTFIELDS, $jsonData);

curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
$response = curl_exec($ch);

curl_close($ch);

echo $response;

function generateHashKey($total, $installment, $currency_code, $merchant_key, $invoice_id, $app_secret)
{
    $data = $total . '|' . $installment . '|' . $currency_code . '|' . $merchant_key . '|' . $invoice_id;

    $iv = substr(sha1(mt_rand()), 0, 16);
    $password = sha1($app_secret);

    $salt = substr(sha1(mt_rand()), 0, 4);
    $saltWithPassword = hash('sha256', $password . $salt);

    $encrypted = openssl_encrypt("$data", 'aes-256-cbc', "$saltWithPassword", null, $iv);

    $msg_encrypted_bundle = "$iv:$salt:$encrypted";
    $msg_encrypted_bundle = str_replace('/', '__', $msg_encrypted_bundle);

    return $msg_encrypted_bundle;
}

function getToken($app_id, $app_secret) {
    $baseUrl = "https://testapp.platformode.com.tr/ccpayment/api/token";

    $data = array(
        'app_id' => $app_id,
        'app_secret' => $app_secret
    );

    $jsonData = json_encode($data);

    $ch = curl_init($baseUrl);

    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $jsonData);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_HTTPHEADER, array(
        'Content-Type: application/json',
    ));

    $response = curl_exec($ch);
	
	$decodedResponse = json_decode($response, true);
	
	return $response;

    curl_close($ch);
}

?>
